### PR TITLE
1110: Add ibm/v1 route features

### PR DESCRIPTION
**Pull request by sunharis**
This commit adds all the usecases of /ibm/v1
    1. ConfigFile multipart update
    2. Lock management fixes
    3. SignCSR feature for VMI interface
    4. Broadcast message event

### CODE REVIEW

**Comment by sunharis** [comment]
@geissonator Can you please help me with this CI failure? What I should do to resolve this ?

**Comment by geissonator** [comment]
Got it running, please take a look at this fail:
```
In file included from /usr/local/include/gtest/gtest-printers.h:122,
                 from /usr/local/include/gtest/gtest-matchers.h:49,
                 from /usr/local/include/gtest/internal/gtest-death-test-internal.h:47,
                 from /usr/local/include/gtest/gtest-death-test.h:43,
                 from /usr/local/include/gtest/gtest.h:65,
                 from ../test/include/ibm/configfile_test.cpp:6:
../test/include/ibm/configfile_test.cpp: In member function ‘virtual void crow::ibm_mc::IsValidConfigFileName_FileNameValidCharReturnsTrue_Test::TestBody()’:
../test/include/ibm/configfile_test.cpp:21:57: error: invalid initialization of reference of type ‘nlohmann::json_abi_v3_11_2::json&’ {aka ‘nlohmann::json_abi_v3_11_2::basic_json<>&’} from expression of type ‘crow::Response’
   21 |     EXPECT_TRUE(isValidConfigFileName("GoodConfigFile", res));
      |                                                         ^~~
In file included from ../test/include/ibm/configfile_test.cpp:2:
../include/ibm/management_console_rest.hpp:62:51: note: in passing argument 2 of ‘bool crow::ibm_mc::isValidConfigFileName(const std::string&, nlohmann::json_abi_v3_11_2::json&)’
   62 |                                   nlohmann::json& resp)
      |                                   ~~~~~~~~~~~~~~~~^~~~
      ```
https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/build/job/github-public/job/repo-ci/job/public-ibm-openbmc-downstream-repo-ci-pr-bmcweb/2158/console

**Comment by sunharis** [comment]
> Got it running, please take a look at this fail:
> 
> ```
> In file included from /usr/local/include/gtest/gtest-printers.h:122,
>                  from /usr/local/include/gtest/gtest-matchers.h:49,
>                  from /usr/local/include/gtest/internal/gtest-death-test-internal.h:47,
>                  from /usr/local/include/gtest/gtest-death-test.h:43,
>                  from /usr/local/include/gtest/gtest.h:65,
>                  from ../test/include/ibm/configfile_test.cpp:6:
> ../test/include/ibm/configfile_test.cpp: In member function ‘virtual void crow::ibm_mc::IsValidConfigFileName_FileNameValidCharReturnsTrue_Test::TestBody()’:
> ../test/include/ibm/configfile_test.cpp:21:57: error: invalid initialization of reference of type ‘nlohmann::json_abi_v3_11_2::json&’ {aka ‘nlohmann::json_abi_v3_11_2::basic_json<>&’} from expression of type ‘crow::Response’
>    21 |     EXPECT_TRUE(isValidConfigFileName("GoodConfigFile", res));
>       |                                                         ^~~
> In file included from ../test/include/ibm/configfile_test.cpp:2:
> ../include/ibm/management_console_rest.hpp:62:51: note: in passing argument 2 of ‘bool crow::ibm_mc::isValidConfigFileName(const std::string&, nlohmann::json_abi_v3_11_2::json&)’
>    62 |                                   nlohmann::json& resp)
>       |                                   ~~~~~~~~~~~~~~~~^~~~
>       ```
> https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/build/job/github-public/job/repo-ci/job/public-ibm-openbmc-downstream-repo-ci-pr-bmcweb/2158/console
> ```

Thank you Andrew. I did not get this error at local devtool ! I will check and fix this.

**Comment by sunharis** [comment]
Jenkins run tests please

**Comment by sunharis** [comment]
Jenkins run tests please

**Comment by sunharis** [comment]
The latest CI failure needs the fixes from PR https://github.com/ibm-openbmc/bmcweb/pull/834

**Comment by gtmills** [comment]
Jenkins run tests please

**Comment by sunharis** [comment]
Jenkins run tests please

File: include/ibm/management_console_rest.hpp [hunk 1]
```diff
@@ -642,38 +851,284 @@ inline void
...
+            strData.append(" ");
+            strData.append(std::to_string(value));
+        }
+
+        asyncResp->res.addHeader("Content-Type", "application/octet-stream");
+        asyncResp->res.write(strData.c_str());
+    };
+
+    crow::connections::systemBus->async_method_call(
+        respHandler, "org.open_power.OCC.Control",
```
Comment (raviteja-b) [inline include/ibm/management_console_rest.hpp:1129]: which feature needs this OCC pass through commands?

**Review by raviteja-b — APPROVED** [review]

File: include/ibm/management_console_rest.hpp (hunk 1 above)
Comment (sunharis) [reply include/ibm/management_console_rest.hpp:1129]: https://github.com/ibm-openbmc/bmcweb/commit/3ba2a2e767acb811c9433ad73bfe217ce44bae9a

**Review by asmithakarun — APPROVED** [review]
LGTM

File: include/ibm/management_console_rest.hpp [hunk 2]
```diff
@@ -32,58 +37,94 @@ constexpr const char* methodNotAllowedMsg = "Method Not Allowed";
...
+    std::error_code ec;
+    // Get the file size getting uploaded
+    BMCWEB_LOG_DEBUG("data length: {}", data.length());
+    BMCWEB_LOG_DEBUG("fileID: {} ", fileID);
+
+    if (data.length() < minSaveareaFileSize)
+    {
+        asyncResp->res.result(boost::beast::http::status::bad_request);
+        asyncResp->res.jsonValue["Description"] =
+            "File size is less than minimum allowed size[100B]";
```
Comment (baemyung) [inline include/ibm/management_console_rest.hpp:104]: It looks like `100B` is defined as minSaveareaFileSize.
Should we write this like?

```
        asyncResp->res.jsonValue["Description"] = std::format(
            "File size is less than minimum allowed size[{}]", minSaveareaFileSize);
```

File: include/ibm/management_console_rest.hpp [hunk 3]
```diff
@@ -32,58 +37,94 @@ constexpr const char* methodNotAllowedMsg = "Method Not Allowed";
...
+        return false;
+    }
+    if (data.length() > maxSaveareaFileSize)
+    {
+        asyncResp->res.result(boost::beast::http::status::bad_request);
+        asyncResp->res.jsonValue["Description"] =
+            "File size exceeds maximum allowed size[25MB]";
```
Comment (baemyung) [inline include/ibm/management_console_rest.hpp:111]: Similarly,
```
        asyncResp->res.jsonValue["Description"] = std::format(
            "File size exceeds maximum allowed size[{}]", maxSaveareaFileSize);
```

File: include/ibm/management_console_rest.hpp [hunk 4]
```diff
@@ -94,47 +135,30 @@ inline void handleFilePut(const crow::Request& req,
...
             {
                 asyncResp->res.result(
                     boost::beast::http::status::internal_server_error);
-                asyncResp->res.jsonValue["Description"] = internalServerError;
-                BMCWEB_LOG_DEBUG("handleIbmPut: Failed to find save-area "
+                asyncResp->res.jsonValue["Description"] =
+                    internalFileSystemError;
+                BMCWEB_LOG_DEBUG("saveConfigFile: Failed to find save-area "
                                  "file size inside the directory . ec : {}",
                                  ec.message());
```
Comment (baemyung) [inline include/ibm/management_console_rest.hpp:154]: Upstream bmcweb suggests to use `ec.value()`.

**Review by baemyung — APPROVED** [review]
I think I may not be able to give the functional comment, but I'm providing a few minor convention related comments.

**Pull request merged**